File: backend/services/inspection_vocabulary.py
"""
Canonical taxonomy for construction inspection terminology.

This is the single source of truth for the controlled vocabulary used to
extract and classify entities out of free-text evidence / inspection notes
(see backend/ai/pipelines/term_extractor.py, which is wired into
inspection_mapping.py).

Mirrored in shared/inspection_vocabulary.ts for the frontend. If you add,
rename, or remove a term here, update the TS mirror in the same PR — see
the contract-sync note at the bottom of this file.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class VocabCategory(str, Enum):
    PROJECT_IDENTIFIER = "project_identifier"
    INSPECTION_TYPE = "inspection_type"
    INSPECTION_STATUS = "inspection_status"
    LOCATION_TERM = "location_term"
    TRADE_TERM = "trade_term"
    DRAWING_TERM = "drawing_term"
    SHEET_IDENTIFIER = "sheet_identifier"
    DOCUMENT_REFERENCE = "document_reference"
    MARKUP_TERM = "markup_term"
    FIELD_CONDITION_TERM = "field_condition_term"
    INSPECTION_ACTION_TERM = "inspection_action_term"
    CONFIDENCE_LABEL = "confidence_label"


class MatchStrategy(str, Enum):
    """How terms in this category should be located in free text."""

    # Exact phrase match, case-insensitive, word-boundary aware. Good for
    # multi-word terms drawn from a closed list ("Hydrostatic Test").
    PHRASE = "phrase"

    # Regex pattern match. Required for structured identifiers like sheet
    # numbers (U1.C4.31) where phrase-matching a finite list is impossible
    # and substring matching would false-positive constantly.
    PATTERN = "pattern"


@dataclass(frozen=True)
class VocabTerm:
    """A single canonical term, plus any alternate surface forms that
    should normalize to it (case/punctuation variants the field notes
    actually use, e.g. "Rough-In", "rough in", "ROUGH IN" -> "Rough In").
    """

    canonical: str
    aliases: tuple[str, ...] = field(default_factory=tuple)


@dataclass(frozen=True)
class VocabCategoryDef:
    category: VocabCategory
    strategy: MatchStrategy
    terms: tuple[VocabTerm, ...] = field(default_factory=tuple)
    # Only used when strategy == PATTERN. Confidence terms also skip
    # matching entirely (see CONFIDENCE_LABEL note below).
    patterns: tuple[str, ...] = field(default_factory=tuple)


def _phrase_terms(*canonicals: str) -> tuple[VocabTerm, ...]:
    return tuple(VocabTerm(canonical=c) for c in canonicals)


# ---------------------------------------------------------------------------
# Taxonomy definition
# ---------------------------------------------------------------------------
# Aliases are seeded conservatively (case/hyphen/spacing variants actually
# seen in field notes). Extend as real-world variants are observed —
# don't guess exhaustively up front.

VOCABULARY: dict[VocabCategory, VocabCategoryDef] = {
    VocabCategory.PROJECT_IDENTIFIER: VocabCategoryDef(
        category=VocabCategory.PROJECT_IDENTIFIER,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Project"),
            VocabTerm("Project Number", aliases=("Project No.", "Project #", "Proj No")),
            VocabTerm("Permit"),
            VocabTerm("Phase"),
            VocabTerm("Area"),
            VocabTerm("Building", aliases=("Bldg",)),
            VocabTerm("Facility"),
            VocabTerm("Campus"),
            VocabTerm("Site"),
            VocabTerm("Utility"),
            VocabTerm("NPC"),
            VocabTerm("UCSF"),
            VocabTerm("Benioff"),
        ),
    ),
    VocabCategory.INSPECTION_TYPE: VocabCategoryDef(
        category=VocabCategory.INSPECTION_TYPE,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Underground Fire Water Rough In", aliases=("UG Fire Water Rough-In",)),
            VocabTerm("Rough In", aliases=("Rough-In", "Rough in")),
            VocabTerm("Final"),
            VocabTerm("Partial Final"),
            VocabTerm("Above Ground", aliases=("Above-Ground", "AG")),
            VocabTerm("Underground", aliases=("Under Ground", "UG")),
            VocabTerm("Hydrostatic Test", aliases=("Hydro Test", "Hydro")),
            VocabTerm("Flush"),
            VocabTerm("Acceptance Test"),
            VocabTerm("Fire Water"),
            VocabTerm("Fire Protection", aliases=("FP",)),
            VocabTerm("Sprinkler"),
            VocabTerm("Underground Piping"),
            VocabTerm("Underground Utilities"),
        ),
    ),
    VocabCategory.INSPECTION_STATUS: VocabCategoryDef(
        category=VocabCategory.INSPECTION_STATUS,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Open"),
            VocabTerm("Closed"),
            VocabTerm("Approved"),
            VocabTerm("Approved As Noted", aliases=("Approved as Noted", "AAN")),
            VocabTerm("Rejected"),
            VocabTerm("Pending"),
            VocabTerm("In Progress"),
            VocabTerm("Scheduled"),
            VocabTerm("Completed"),
            VocabTerm("Passed"),
            VocabTerm("Failed"),
            VocabTerm("Deferred"),
        ),
    ),
    VocabCategory.LOCATION_TERM: VocabCategoryDef(
        category=VocabCategory.LOCATION_TERM,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Utility MR", aliases=("Utility Mechanical Room",)),
            VocabTerm("Mechanical Room", aliases=("Mech Room", "MR")),
            VocabTerm("Equipment Room"),
            VocabTerm("Site"),
            VocabTerm("Yard"),
            VocabTerm("Corridor"),
            VocabTerm("Level"),
            VocabTerm("Floor"),
            VocabTerm("Roof"),
            VocabTerm("Exterior"),
            VocabTerm("Interior"),
            VocabTerm("Building Area"),
            VocabTerm("Grid Line"),
            VocabTerm("Coordinate"),
            VocabTerm("Utility Corridor"),
        ),
    ),
    VocabCategory.TRADE_TERM: VocabCategoryDef(
        category=VocabCategory.TRADE_TERM,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Fire Protection", aliases=("FP",)),
            VocabTerm("Mechanical"),
            VocabTerm("Plumbing"),
            VocabTerm("Electrical"),
            VocabTerm("Structural"),
            VocabTerm("Civil"),
            VocabTerm("Architectural"),
            VocabTerm("Underground Utilities"),
        ),
    ),
    VocabCategory.DRAWING_TERM: VocabCategoryDef(
        category=VocabCategory.DRAWING_TERM,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Drawing"),
            VocabTerm("Sheet"),
            VocabTerm("Sheet Number", aliases=("Sheet No.", "Sheet #")),
            VocabTerm("Plan"),
            VocabTerm("Detail"),
            VocabTerm("Section"),
            VocabTerm("Elevation"),
            VocabTerm("Revision", aliases=("Rev",)),
            VocabTerm("Attachment"),
            VocabTerm("Reference Drawing"),
            VocabTerm("Master Drawing"),
            VocabTerm("Construction Drawing"),
        ),
    ),
    VocabCategory.SHEET_IDENTIFIER: VocabCategoryDef(
        category=VocabCategory.SHEET_IDENTIFIER,
        strategy=MatchStrategy.PATTERN,
        # Examples from the audit: U1.C4.31, U1.C4.32, C4.31, C4.32
        # General shape: optional 1-3 alnum prefix block + dot-separated
        # alnum segments, at least 2 segments, each segment 1-3 chars.
        # Anchored to avoid matching arbitrary decimal numbers like "4.31".
        patterns=(
            r"\b[A-Z]{1,3}\d{0,2}(?:\.[A-Z0-9]{1,4}){1,4}\b",
        ),
    ),
    VocabCategory.DOCUMENT_REFERENCE: VocabCategoryDef(
        category=VocabCategory.DOCUMENT_REFERENCE,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Attachment"),
            VocabTerm("Linked Drawing"),
            VocabTerm("Referenced Drawing"),
            VocabTerm("Supporting Document"),
            VocabTerm("Inspection Package"),
            VocabTerm("Record Drawing"),
        ),
    ),
    VocabCategory.MARKUP_TERM: VocabCategoryDef(
        category=VocabCategory.MARKUP_TERM,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Cloud"),
            VocabTerm("Revision Cloud"),
            VocabTerm("Arrow"),
            VocabTerm("Leader"),
            VocabTerm("Callout"),
            VocabTerm("Note"),
            VocabTerm("Comment"),
            VocabTerm("Markup"),
            VocabTerm("Highlight"),
            VocabTerm("Stamp"),
            VocabTerm("Inspection Tag"),
            VocabTerm("Deficiency Marker"),
        ),
    ),
    VocabCategory.FIELD_CONDITION_TERM: VocabCategoryDef(
        category=VocabCategory.FIELD_CONDITION_TERM,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Installed"),
            VocabTerm("Existing", aliases=("Exist.", "Exist")),
            VocabTerm("New"),
            VocabTerm("Verify"),
            VocabTerm("Confirm"),
            VocabTerm("Relocate"),
            VocabTerm("Remove"),
            VocabTerm("Modify"),
            VocabTerm("Correct"),
            VocabTerm("Repair"),
            VocabTerm("Replace"),
        ),
    ),
    VocabCategory.INSPECTION_ACTION_TERM: VocabCategoryDef(
        category=VocabCategory.INSPECTION_ACTION_TERM,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("Inspect"),
            VocabTerm("Verify"),
            VocabTerm("Observe"),
            VocabTerm("Witness"),
            VocabTerm("Review"),
            VocabTerm("Test"),
            VocabTerm("Approve"),
            VocabTerm("Reject"),
            VocabTerm("Close"),
            VocabTerm("Document"),
        ),
    ),
    VocabCategory.CONFIDENCE_LABEL: VocabCategoryDef(
        category=VocabCategory.CONFIDENCE_LABEL,
        strategy=MatchStrategy.PHRASE,
        terms=(
            VocabTerm("High Confidence"),
            VocabTerm("Medium Confidence"),
            VocabTerm("Low Confidence"),
        ),
    ),
}


# ---------------------------------------------------------------------------
# Convenience accessors
# ---------------------------------------------------------------------------

def all_categories() -> tuple[VocabCategory, ...]:
    return tuple(VOCABULARY.keys())


def category_def(category: VocabCategory) -> VocabCategoryDef:
    return VOCABULARY[category]


def canonical_terms(category: VocabCategory) -> tuple[str, ...]:
    """Canonical term strings for a category (PHRASE categories only)."""
    return tuple(t.canonical for t in VOCABULARY[category].terms)


# ---------------------------------------------------------------------------
# Contract-sync note
# ---------------------------------------------------------------------------
# shared/inspection_vocabulary.ts mirrors this file's category names and
# canonical terms (not aliases or regex patterns — those are extraction
# implementation detail the frontend doesn't need). When this file changes
# in a way that affects category names or canonical terms, update the TS
# mirror in the same PR, the same way shared/schema.ts is kept in sync with
# backend/models/schemas.py elsewhere in this refactor.

File: shared/inspection_vocabulary.ts
/**
 * TypeScript mirror of backend/services/inspection_vocabulary.py.
 *
 * This mirrors category names and CANONICAL terms only — not aliases or
 * extraction regex patterns, since those are backend extraction
 * implementation detail the frontend doesn't need. The frontend's job is
 * to render already-normalized tags (from DrawingOverlay / inspection_runs
 * records) using consistent labels, category colors, and grouping — not
 * to re-run extraction itself.
 *
 * Contract-sync rule: if backend/services/inspection_vocabulary.py changes
 * a category name or a canonical term, update this file in the same PR —
 * the same convention used for shared/schema.ts <-> backend/models/schemas.py
 * elsewhere in this refactor.
 */

export enum VocabCategory {
  ProjectIdentifier = "project_identifier",
  InspectionType = "inspection_type",
  InspectionStatus = "inspection_status",
  LocationTerm = "location_term",
  TradeTerm = "trade_term",
  DrawingTerm = "drawing_term",
  SheetIdentifier = "sheet_identifier",
  DocumentReference = "document_reference",
  MarkupTerm = "markup_term",
  FieldConditionTerm = "field_condition_term",
  InspectionActionTerm = "inspection_action_term",
  ConfidenceLabel = "confidence_label",
}

export enum ConfidenceLabel {
  High = "High Confidence",
  Medium = "Medium Confidence",
  Low = "Low Confidence",
}

/** Canonical terms per category. Keep in the same order as the Python
 * source for easy diffing. SheetIdentifier has no fixed list (pattern-
 * matched on the backend) so it's intentionally empty here. */
export const VOCABULARY: Record<VocabCategory, readonly string[]> = {
  [VocabCategory.ProjectIdentifier]: [
    "Project",
    "Project Number",
    "Permit",
    "Phase",
    "Area",
    "Building",
    "Facility",
    "Campus",
    "Site",
    "Utility",
    "NPC",
    "UCSF",
    "Benioff",
  ],
  [VocabCategory.InspectionType]: [
    "Underground Fire Water Rough In",
    "Rough In",
    "Final",
    "Partial Final",
    "Above Ground",
    "Underground",
    "Hydrostatic Test",
    "Flush",
    "Acceptance Test",
    "Fire Water",
    "Fire Protection",
    "Sprinkler",
    "Underground Piping",
    "Underground Utilities",
  ],
  [VocabCategory.InspectionStatus]: [
    "Open",
    "Closed",
    "Approved",
    "Approved As Noted",
    "Rejected",
    "Pending",
    "In Progress",
    "Scheduled",
    "Completed",
    "Passed",
    "Failed",
    "Deferred",
  ],
  [VocabCategory.LocationTerm]: [
    "Utility MR",
    "Mechanical Room",
    "Equipment Room",
    "Site",
    "Yard",
    "Corridor",
    "Level",
    "Floor",
    "Roof",
    "Exterior",
    "Interior",
    "Building Area",
    "Grid Line",
    "Coordinate",
    "Utility Corridor",
  ],
  [VocabCategory.TradeTerm]: [
    "Fire Protection",
    "Mechanical",
    "Plumbing",
    "Electrical",
    "Structural",
    "Civil",
    "Architectural",
    "Underground Utilities",
  ],
  [VocabCategory.DrawingTerm]: [
    "Drawing",
    "Sheet",
    "Sheet Number",
    "Plan",
    "Detail",
    "Section",
    "Elevation",
    "Revision",
    "Attachment",
    "Reference Drawing",
    "Master Drawing",
    "Construction Drawing",
  ],
  [VocabCategory.SheetIdentifier]: [],
  [VocabCategory.DocumentReference]: [
    "Attachment",
    "Linked Drawing",
    "Referenced Drawing",
    "Supporting Document",
    "Inspection Package",
    "Record Drawing",
  ],
  [VocabCategory.MarkupTerm]: [
    "Cloud",
    "Revision Cloud",
    "Arrow",
    "Leader",
    "Callout",
    "Note",
    "Comment",
    "Markup",
    "Highlight",
    "Stamp",
    "Inspection Tag",
    "Deficiency Marker",
  ],
  [VocabCategory.FieldConditionTerm]: [
    "Installed",
    "Existing",
    "New",
    "Verify",
    "Confirm",
    "Relocate",
    "Remove",
    "Modify",
    "Correct",
    "Repair",
    "Replace",
  ],
  [VocabCategory.InspectionActionTerm]: [
    "Inspect",
    "Verify",
    "Observe",
    "Witness",
    "Review",
    "Test",
    "Approve",
    "Reject",
    "Close",
    "Document",
  ],
  [VocabCategory.ConfidenceLabel]: [
    ConfidenceLabel.High,
    ConfidenceLabel.Medium,
    ConfidenceLabel.Low,
  ],
};

/** The shape of a single extracted/normalized term, as produced by
 * backend.ai.pipelines.term_extractor.ExtractedTerm.to_dict() and
 * returned over the wire. Field names are camelCase to match the
 * to_dict() serialization. */
export interface ExtractedTerm {
  category: VocabCategory;
  canonical: string;
  matchedText: string;
  start: number;
  end: number;
  confidenceScore: number;
  confidenceLabel: ConfidenceLabel;
}

/** Mirrors NormalizedEvidenceTags.to_dict() from inspection_mapping.py —
 * the normalized-tag bundle attached to a DrawingOverlay / inspection
 * finding once evidence text has been run through extraction. */
export interface NormalizedEvidenceTags {
  inspectionTypes: string[];
  inspectionStatuses: string[];
  locations: string[];
  trades: string[];
  fieldConditions: string[];
  actions: string[];
  markupTerms: string[];
  confidenceLabel: ConfidenceLabel;
}

/** Convenience lookup for UI components that need every canonical term
 * for a category, e.g. populating a filter dropdown on the inspection
 * runs sidebar. */
export function canonicalTermsFor(category: VocabCategory): readonly string[] {
  return VOCABULARY[category];
}

/** Display color tokens per category, for consistent overlay/tag styling
 * across the inspection runs sidebar, overlay layer, and findings list.
 * Adjust to the app's actual design tokens — these are placeholders. */
export const VOCAB_CATEGORY_COLOR: Record<VocabCategory, string> = {
  [VocabCategory.ProjectIdentifier]: "#64748B",
  [VocabCategory.InspectionType]: "#2E5395",
  [VocabCategory.InspectionStatus]: "#1F8A4C",
  [VocabCategory.LocationTerm]: "#B97034",
  [VocabCategory.TradeTerm]: "#7B5EA7",
  [VocabCategory.DrawingTerm]: "#44546A",
  [VocabCategory.SheetIdentifier]: "#0E7C86",
  [VocabCategory.DocumentReference]: "#8A6D3B",
  [VocabCategory.MarkupTerm]: "#C0392B",
  [VocabCategory.FieldConditionTerm]: "#D68910",
  [VocabCategory.InspectionActionTerm]: "#117864",
  [VocabCategory.ConfidenceLabel]: "#566573",
};

/** Maps a confidence label to a severity-style indicator, for surfacing
 * "this extraction needs a human glance" in the sidebar/overlay UI. */
export function confidenceIndicator(
  label: ConfidenceLabel
): "success" | "warning" | "danger" {
  switch (label) {
    case ConfidenceLabel.High:
      return "success";
    case ConfidenceLabel.Medium:
      return "warning";
    case ConfidenceLabel.Low:
      return "danger";
  }
}
